Reorganize queries

jayrbolton opened the pull request:
* Rename "views" to "stored queries"
* Move the AQL into the yaml key "query"
* Add "name" (query name) and "params" (json schema on query parameters) for each query
* Update test/validate.py to work with the above changes

MrCreosote commented on test/validate.py:
```diff
@@ -68,21 +68,43 @@ def validate_json_schemas():
             exit(1)
         else:
             names[name] = True
+        # Make sure `params` can be used as a JSON schema
```
I'm guessing there's no easy way to check that the params match the `@` variables in the query?

jayrbolton replied:
We could pass the query to these endpoints:
https://www.arangodb.com/docs/stable/http/aql-query.html
We can send the query to /_api/query and it'll validate the syntax and send back the bind vars

As long as arangodb is running as part of the test runner (I believe it is), then this should work fine. I'll add this in.

MrCreosote commented on test/validate.py:
```diff
@@ -68,21 +68,43 @@ def validate_json_schemas():
+        if data.get('params'):
+            try:
+                jsonschema.validate({}, data['params'])
+            except ValidationError:
```
Wait, if it fails to validate that's a pass...?

jayrbolton replied:
All this is doing is validating that data['params'] is a valid json schema. If it fails to parse as a json schema, it will throw a SchemaError. If it's a valid json schema, then a ValidationError will get thrown on the empty object

MrCreosote replied:
Ok, you should add a comment for that as it's pretty confusing

jayrbolton commented:
@MrCreosote I added AQL syntax checking, plus bind var validation using the arangodb query parsing endpoint. Turned out to be pretty easy and uncovered a couple bugs

**Note on the @ws_ids bind var**: this is a special bind var that is set by the RE API for every query of all workspaces the requester has access to. In AQL, if you have a bind var set, but it doesn't appear in the query, then an error gets thrown. However, we don't want to use this variable in all queries (eg. taxonomy). So the RE API will prepend the clause `LET ws_ids = @ws_ids` to prevent the error. To be consistent, we can just use `ws_ids` when we need to use that var.

MrCreosote commented:
this is ready for re-review then?

jayrbolton commented:
Ready for re review

On Mon, Jul 22, 2019 at 5:41 PM MrCreosote <notifications@github.com> wrote:

> this is ready for re-review then?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kbase/relation_engine_spec/pull/67?email_source=notifications&email_token=ACCV47FIPR3XHWIOKJOZTEDQAZHVBA5CNFSM4IF5XFP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2RSIDQ#issuecomment-514008078>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACCV47DZU5W6MUR7E7JTBB3QAZHVBANCNFSM4IF5XFPQ>
> .
>

MrCreosote commented on test/helpers.py:
```diff
@@ -0,0 +1,32 @@
+import functools
+
+
+@functools.lru_cache(maxsize=1)
```
Out of curiosity, why is this being cached?

MrCreosote commented on test/helpers.py:
```diff
@@ -0,0 +1,32 @@
+def get_config():
+    """Return configuration data for tests."""
+    return {
+        'db_url': 'http://arangodb:8529',
```
Does the url need to be configurable?

MrCreosote commented on test/helpers.py:
```diff
@@ -0,0 +1,32 @@
+
+def wait_for_arangodb():
+    """Wait for arangodb to go live."""
+    db_url = 'http://arangodb:8529'
```
Same ?

MrCreosote commented on test/validate.py:
```diff
@@ -68,21 +77,65 @@ def validate_json_schemas():
+            print(f'Duplicate queries named {name}')
             exit(1)
         else:
             names[name] = True
```
doesn't really matter, but this could be a set

jayrbolton commented:
@MrCreosote I made some more updates, ready for re-review if you have time:
* Changed the `names` dicts to sets as suggested
* Moved some config into the environment / docker-compose. Makes a little more sense for that config to live there, but it doesn't need to be configurable for this repo. 
* I updated many of requirements and default values for the stored query params based on what i think makes sense with the query. Not confident this is all backwards compatible so I'll have to manually test a bunch of things

For json schema default values, I'm going to add something like this in the RE API: https://python-jsonschema.readthedocs.io/en/stable/faq/

MrCreosote commented:
Looks ok to me other than the question about the cached fn call. I didn't look at the stored queries though.
